atomic_reactor/utils/resource_manager.py

```diff
@@ -270,45 +289,63 @@ def from_config(cls, config: dict):
-                 for hostname, attr in config.items() if attr['enabled']]
+        hosts = []
+        for hostname, attr in config.items():
+            if not attr.get("enabled", False):
+                continue
+            host = RemoteHost(
+                hostname=hostname, username=attr["username"], ssh_keyfile=attr["auth"],
+                slots=attr.get("slots", 1)
+            )
+            # Check whether host is operational before use it
+            if host.is_operational:
+                hosts.append(host)
 
         return cls(hosts)
 
-    @backoff.on_exception(
-        backoff.constant,
-        LockRetry,
-        max_time=TIMEOUT_FAIL_BUILD,
-        jitter=None,
-    )
-    def lock_resource(self, prid: str) -> LockedResource:
-        """ Lock resources for build """
+    def lock_resource(self, prid: str) -> Optional[LockedResource]:
+        """
+        Lock resource for a pipelinerun
+
+        :param prid: str, pipelinerun ID
+        """
+
+        if not self.hosts:
+            logger.error("This is no available remote host in pool")
+            return None
 
-        # Slots should be randomized here to avoid always locking the lowest numbers first
         resources = []
         for host in self.hosts:
             available_slots = host.available_slots()
-            random.shuffle(available_slots)
+            if not available_slots:
+                logger.info("%s: no available slots", host.hostname)
+                continue
+            logger.info("%s: available slots: %s", host.hostname, available_slots)
             resources.append((host, available_slots))
 
-        # Sort list based on number of available slots
-        resources.sort(key=lambda x: len(x[1]), reverse=True)
+        if not resources:
+            logger.error("There is no remote host slot available for pipelinerun %s", prid)
+            return None
+
+        # Sort list based on ratio of available_slots/all_slots
+        resources.sort(key=lambda x: len(x[1])/x[0].slots, reverse=True)
 
-        # Try to lock resources
+        # Try to lock a remote host slot for pipelinerun
         for host, slots in resources:
             for slot in slots:
                 try:
                     host.lock(slot, prid)
-                except Exception as ex:
+                except Exception as e:
                     # Specific exceptions should be handled in nested methods
-                    logger.warning("Locking failed on slot %s, host %s - %s", slot, host, ex)
+                    logger.warning("%s: unable to lock slot %s for pipelinerun %s: %s",
+                                   host.hostname, slot, prid, e)
                 else:
                     return LockedResource(host, slot, prid)
 
-        raise LockRetry(f'No remote host resources were available for pipelinerun ID {prid}')
+        logger.info("Cannot find remote host resource for pipelinerun %s", prid)
+        return None
```

Comment:
Do we not want to retry the `lock_resource()` operation anymore? Or do you think it should be up to the code that uses this module?

Reply:
Not sure, I'll revisit this and do it in future patch after #1808 is merged.

Reply:
Not sure, I'll revisit this and do it in future patch after #1808 is merged.